test/symforce_rust_codegen_test.py

```diff
@@ -0,0 +1,144 @@
+# ----------------------------------------------------------------------------
+# SymForce - Copyright 2022, Skydio, Inc.
+# This source code is under the Apache 2.0 license found in the LICENSE file.
+# ----------------------------------------------------------------------------
+
+import symforce
+
+symforce.set_epsilon_to_symbol()
+
+import functools
+from pathlib import Path
+import shutil
+import subprocess
+import textwrap
+
+import symforce.symbolic as sf
+from symforce import path_util
+from symforce import typing as T
+from symforce.codegen import Codegen
+from symforce.codegen import CodeGenerationException
+from symforce.codegen.backends.rust import RustConfig
+from symforce.codegen.backends.rust import ScalarType
+from symforce.test_util import TestCase
+from symforce.test_util.backend_coverage_expressions import backend_test_function
+
+TEST_DATA_DIR = (
+    path_util.symforce_data_root()
+    / "test"
+    / "symforce_function_codegen_test_data"
+    / symforce.get_symbolic_api()
+    / "symforce_rust_codegen_test"
+)
+
+
+def write_cargo_toml(output_dir: Path) -> None:
+    """Write the Cargo.toml file for the generated code."""
+    (output_dir / "Cargo.toml").write_text(
+        textwrap.dedent(
+            """
+            [package]
+            name = "symforce_rust_codegen_test"
+            version = "0.1.0"
+            edition = "2021"
+
+            [dependencies]
+            nalgebra = "0"
+            """
+        )
+    )
+
+
+
+
+def write_lib_rs(output_dir: Path) -> None:
+    """Write the lib.rs file that includes the generated functions."""
+    (output_dir / "lib.rs").write_text(
+        textwrap.dedent(
+            """
+            mod backend_test_function_float32;
+            mod backend_test_function_float64;
+            mod vector_matrix_fun;
+            """
+        )
+    )
+
+def cargo_build(output_dir: Path) -> None:
+    """Run cargo build in the output directory
+
+    Note that this fails when called from within a temporary directory!
+    """
+    result = subprocess.run(
+        ["cargo", "build", "--manifest-path", output_dir / "Cargo.toml"],
+        capture_output=True,
+        text=True,
+    )
+
+    # Always clean up after ourselves even if the build fails.
+    cargo_clean(output_dir)
+
+    if result.returncode != 0:
+        assert False, f"cargo build failed:\n{result.stderr}"
```

Comment:
Can we call `self.fail` here (with `self` being the `TestCase`) instead of asserting?

Reply:
Done